larray/core/session.py

```diff
@@ -1393,87 +1394,85 @@ def display(k, v, is_metadata=False):
-    filepath: str, optional
-        Path where items have been saved. This can be either the path to a single file, a path to
-        a directory containing .csv files or a pattern representing several .csv files.
-    meta : list of pairs or dict or OrderedDict or Metadata, optional
-        Metadata (title, description, author, creation_date, ...) associated with the array.
-        Keys must be strings. Values must be of type string, int, float, date, time or datetime.
-
-    Notes
-    -----
-    The 'autocomplete' is a feature in which a software predicts the rest of a variable or function
-    name after a user typed the first letters. This feature allows users to use longer but meaningful
-    variable or function names (like 'population_be' instead of 'pbe') and to avoid creating an unwanted
-    new variable by misspelling the name of a given variable (e.g. typing 'poplation = something'
-    ('population' without u) will create a new variable instead of modifying the 'population' variable).
 
+# XXX: the name of the class below is not really important since it will serve as base
+#      for the LazySession and users are expected to inherit from the LazySession when
+#      they define their own classes
+class TypedSession(Session):
+    """
     Examples
     --------
-    >>> class ModelVariables(FrozenSession):
-    ...     LAST_AGE = 120
-    ...     FIRST_OBS_YEAR = 1991
-    ...     LAST_PROJ_YEAR = 2070
-    ...     AGE = Axis('age=0..{}'.format(LAST_AGE))
-    ...     GENDER = Axis('gender=male,female')
-    ...     TIME = Axis('time={}..{}'.format(FIRST_OBS_YEAR, LAST_PROJ_YEAR))
-    ...     CHILDREN = AGE[0:17]
-    ...     ELDERS = AGE[65:]
-    ...     population = zeros((AGE, GENDER, TIME))
-    ...     births = zeros((AGE, GENDER, TIME))
-    ...     deaths = zeros((AGE, GENDER, TIME))
-    >>> m = ModelVariables()
-    >>> m.names    # doctest: +NORMALIZE_WHITESPACE
-    ['AGE', 'CHILDREN', 'ELDERS', 'FIRST_OBS_YEAR', 'GENDER', 'LAST_AGE', 'LAST_PROJ_YEAR', 'TIME', 'births',
-    'deaths', 'population']
+    Content of file 'model_variables.py'
+
+    >>> # ==== MODEL VARIABLES ====
+    >>> class ModelVariables(TypedSession):
+    ...     FIRST_OBS_YEAR = int
+    ...     FIRST_PROJ_YEAR = int
+    ...     LAST_PROJ_YEAR = int
+    ...     AGE = Axis
+    ...     GENDER = Axis
+    ...     TIME = Axis
+    ...     G_CHILDREN = Group
+    ...     G_ADULTS = Group
+    ...     G_OBS_YEARS = Group
+    ...     G_PROJ_YEARS = Group
+    ...     population = ArrayDef(('AGE', 'GENDER', 'TIME'))
+    ...     births = ArrayDef(('AGE', 'GENDER', 'TIME'))
+    ...     deaths = ArrayDef(('AGE', 'GENDER', 'TIME'))
+
+    Content of file 'model.py'
+
+    >>> def run_model(variant_name, first_proj_year, last_proj_year):
+    ...     # create an instance of the ModelVariables class
+    ...     m = ModelVariables()
+    ...     # ==== setup variables ====
+    ...     # set scalars
+    ...     m.FIRST_OBS_YEAR = 1991
```

Comment:
Right. 

But there is something else for what I would like some advices. The main drawbacks of allowing to **refer to axes defined at runtime** in ``var = ArrayDef((...))`` are:

1.   they must be referred using strings -> we loose the benefit auto-completion + make it dangerous to rename these axes (via refactoring (Shift + F6) in PyCharm)
2.   the following code:
```python
m = ModelVariables()
....
# save to Excel file -> save Array objects only (not Axis objects)
m.save('variant_0.xlsx')
...
# load variables for variant 0 from Excel file -> FAIL
m = ModelVariables('variant_0.xlsx')
```
will fail because Axis objects are not saved in Excel or CSV files.

what's your opinion on this?

Reply:
IMO, this is not a problem. If they want to do this, they should not use an Excel file! Now, it's up to us to make this clear in the documentation. In the code, nothing needs to change I think as there is already a warning that the axis object is not saved.